File: day03/day03.jl
# Day 03: Gear Ratios

# ==============================================================================

# Get input filename as CLI arg. If not given, default to "day<X>.in"
if length(ARGS) > 0
  fname = ARGS[1]
else
  fname = replace(PROGRAM_FILE, ".jl" => ".in")
end

# Parse input
lines = readlines(fname)
nx = length(lines)
ny = length(lines[1])

# ------------------------------------------------------------------------------
# Part 1

function is_part_number(l, x1, x2)
  for i in l-1:l+1
    for j in x1-1:x2+1
      if 1 <= i <= nx && 1 <= j <= ny
        if !isdigit(lines[i][j]) && lines[i][j] != '.'
          return true
        end
      end
    end
  end
  return false
end

total = 0
numbers = Dict()
for l in 1:length(lines)
  line = lines[l]
  x = 1
  while x <= nx
    if isdigit(line[x])
      x1 = x
      while x1+1 <= nx && isdigit(line[x1+1])
        x1 += 1
      end
      number = parse(Int, line[x:x1])
      if !haskey(numbers, l)
        numbers[l] = []
      end
      push!(numbers[l], (number, x, x1))
      if is_part_number(l, x, x1)
        global total += number
      end
      x = x1
    end 
    x += 1
  end 
end


println("Part 1: ", total)

# ------------------------------------------------------------------------------
# Part 2

function find_adjacent_numbers(x, y)
  adjacent = []
  for xx in (x-1, x, x+1)
    if 1 <= xx <= nx
      for (num, j1, j2) in numbers[xx]
        if j1-1 <= y <= j2+1
          push!(adjacent, num)
        end
      end
    end
  end
  return adjacent
end

total = 0
for x in 1:nx
  for y in 1:ny
    if lines[x][y] == '*'
      adjacent = find_adjacent_numbers(x, y)
      if length(adjacent) == 2
        gear_ratio = adjacent[1] * adjacent[2]
        global total += gear_ratio
      end
    end
  end
end

println("Part 2: ", total)

File: day03/day03.py
# Day 03: Gear Ratios

import string
import sys

# ==============================================================================

# If no specific input given, default to "day<X>.in"
if len(sys.argv) == 1:
  input_fname = sys.argv[0].replace(".py", ".in")
else:
  input_fname = sys.argv[1]

# Read and parse input
grid = []
with open(input_fname) as f:
  for line in f:
    grid.append(line.strip())
nx = len(grid)
ny = len(grid[0])

# ------------------------------------------------------------------------------
# Part 1

def is_part_number(x1, x2, y1, y2):
  x1 = max(x1, 0)
  x2 = min(x2, nx-1)
  y1 = max(y1, 0)
  y2 = min(y2, ny-1)
  chars = string.digits + "."
  for i in range(x1, x2+1):
    for j in range(y1, y2+1):
      if grid[i][j] not in chars:
        return True
  return False 

numbers = {x:[] for x in range(nx)}
total = 0
x = 0
while x < nx:
  y = 0
  while y < ny:
    if grid[x][y] in string.digits:
      j = y
      while j < ny-1 and grid[x][j+1] in string.digits:
        j += 1
      number = int(grid[x][y:j+1])
      numbers[x].append((y,j,number))
      valid = is_part_number(x-1, x+1, y-1, j+1)
      if valid:
        total += number
      y = j+1
    y += 1
  x += 1

print("Part 1:", total)

# ------------------------------------------------------------------------------
# Part 2

def find_adjacent_numbers(x, y):
  adjacent = []
  for xx in (x-1, x, x+1):
    if 0 <= xx < nx:
      for j1, j2, num in numbers[xx]:
        if j1-1 <= y <= j2+1:
          adjacent.append(num)
  return adjacent

total = 0
x = 0
while x < nx:
  y = 0
  while y < ny:
    if grid[x][y] == "*":
      adjacent = find_adjacent_numbers(x, y)
      if len(adjacent) == 2:
        gear_ratio = adjacent[0] * adjacent[1]
        total += gear_ratio
    y += 1
  x += 1    

print("Part 2:", total)
